sockeye/train.py

```diff
@@ -132,278 +139,446 @@ def main():
     else:
         os.makedirs(output_folder)
 
+    return resume_training, training_state_dir
+
+
+def determine_context(args: argparse.Namespace, exit_stack: ExitStack) -> List[mx.Context]:
+    """
+    Determine the context we should run on (CPU or GPU).
+
+    :param args: Arguments as returned by argparse.
+    :param exit_stack: An ExitStack from contextlib.
+    :return: A list with the context(s) to run on.
+    """
+    if args.use_cpu:
+        logger.info("Device: CPU")
+        context = [mx.cpu()]
+    else:
+        num_gpus = utils.get_num_gpus()
+        check_condition(num_gpus >= 1,
+                        "No GPUs found, consider running on the CPU with --use-cpu "
+                        "(note: check depends on nvidia-smi and this could also mean that the nvidia-smi "
+                        "binary isn't on the path).")
+        if args.disable_device_locking:
+            context = utils.expand_requested_device_ids(args.device_ids)
+        else:
+            context = exit_stack.enter_context(utils.acquire_gpus(args.device_ids, lock_dir=args.lock_dir))
+        logger.info("Device(s): GPU %s", context)
+        context = [mx.gpu(gpu_id) for gpu_id in context]
+    return context
+
+
+def load_or_create_vocabs(args: argparse.Namespace, resume_training: bool, output_folder: str) -> Tuple[Dict, Dict]:
+    """
+    Load the vocabularies from disks if given, create them if not.
+
+    :param args: Arguments as returned by argparse.
+    :param resume_training: Gives if an interrupted training should be resumed.
```

Comment:
"Gives if" -> "When True the vocabulary will be loaded from an existing output folder."